Add support for Word's images pasting to Safari

Comandeer opened the pull request:
<!--
🚨 If you want to submit a PR for a security vulnerability, please contact us directly
at https://ckeditor.com/contact/ instead. 🚨
-->
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#1134](https://github.com/ckeditor/ckeditor4/issues/1134): [Safari] Fixed: [Paste from Word](https://ckeditor.com/cke4/addon/pastefromword) does not embed images.
```

## What changes did you make?

I've basically taken the contents from @msamsel's #1976 and integrated it with our `CKEDITOR.plugins.pastetools.filters.image` filter.

## Which issues does your PR resolve?

Closes #1134.
<!-- Closes #<ANOTHER_ISSUE_NUMBER>. -->

jacekbogdanski commented on plugins/pastetools/filter/image.js:
```diff
@@ -300,13 +362,60 @@
+			header += bytesHeader[ i ].toString( 16 );
+		}
+
+		switch ( header ) {
```
That's some dark magic 😄 I'm not sure if it should be named `header`, probably [File signature](https://en.wikipedia.org/wiki/File_signature) is more accurate here. It would be good to have real, existing images with these signatures, to confirm that the feature covers all expected image types. These can be checked e.g. here: https://www.filesignatures.net/index.php?search=FFD8FFE0&mode=SIG although I suppose it may be hard to get all `spf`, `jfif` etc extension examples and I'm not sure if Word is not converting them also into a more unified format.

jacekbogdanski commented on plugins/ajax/plugin.js:
```diff
@@ -128,12 +136,16 @@
-			 * @returns {String} The loaded data. For asynchronous requests, an
+			 * synchronously. Please notice that only text data might be loaded synchrnously.
+			 * @param {String} [responseType='text'] Defines type of returned data.
+			 * Currently supports: `text`, `xml`, `arraybuffer`. This parameter was added in 4.10.
```
```suggestion
			 * Currently supports: `text`, `xml`, `arraybuffer`. This parameter was added in 4.16.0.
```

jacekbogdanski commented on plugins/ajax/plugin.js:
```diff
@@ -128,12 +136,16 @@
 			 * data load. If not provided, the data will be loaded
-			 * synchronously.
-			 * @returns {String} The loaded data. For asynchronous requests, an
+			 * synchronously. Please notice that only text data might be loaded synchrnously.
```
```suggestion
			 * synchronously. Please notice that only text data might be loaded synchronously.
```

jacekbogdanski commented on tests/plugins/ajax/ajax.js:
```diff
@@ -179,6 +179,27 @@
+		},
+
+		// (#1134)
+		'test load async arraybuffer': function() {
```
We are also missing `text` and `xml` types. I guess they are probably tested already as we are using them indirectly in ajax logic, but it would be good to check if API works correctly.

jacekbogdanski commented on tests/plugins/pastefromword/_helpers/blob.js:
```diff
@@ -0,0 +1,100 @@
+
+	function simulatePasteBlob( editor, assertion, options ) {
+		// jscs:disable maximumLineLength
+		var imgBase64 = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAIAAAACACAYAAADDPmHLAAAEDWlDQ1BJQ0MgUHJvZmlsZQAAOI2NVV1oHFUUPrtzZyMkzlNsNIV0qD8NJQ2TVjShtLp/3d02bpZJNtoi6GT27s6Yyc44M7v9oU9FUHwx6psUxL+3gCAo9Q/bPrQvlQol2tQgKD60+INQ6Ium65k7M5lpurHeZe58853vnnvuuWfvBei5qliWkRQBFpquLRcy4nOHj4g9K5CEh6AXBqFXUR0rXalMAjZPC3e1W99Dwntf2dXd/p+tt0YdFSBxH2Kz5qgLiI8B8KdVy3YBevqRHz/qWh72Yui3MUDEL3q44WPXw3M+fo1pZu ...
```
Do you think it would be possible to use real images instead with more test cases?

jacekbogdanski commented on tests/plugins/pastefromword/manual/pasteimagessafari.html:
```diff
@@ -0,0 +1,7 @@
+<textarea id="editor" cols="30" rows="10"></textarea>
+<script>
+	if ( !CKEDITOR.env.safari ) {
+		bender.ignore();
+	}
```
Maybe it makes sense to keep this test also for different browsers to make sure we didn't introduce regression?

Comandeer replied:
All other browsers are covered by unit tests for Paste Tools, Paste from Word and Paste from Libre Office, so I don't think it would be necessary.

jacekbogdanski commented on tests/plugins/pastefromword/manual/pasteimagessafari.md:
```diff
@@ -0,0 +1,20 @@
+@bender-tags: bug, 4.16.0, 1134
+@bender-ui: collapsed
+@bender-ckeditor-plugins: wysiwygarea, toolbar, undo, basicstyles, pastefromword, elementspath, image, format, sourcearea
+
+1. Open [attached Word document](_assets/Image_safari.docx).
```
It would be good to add a more complicated document with more image sources, especially gif and different flavors of jpg.

Comandeer replied:
How can I produce different flavors of JPG? :thinking:

Comandeer replied:
Actually we probably don't need to test different flavors of JPG, as all of its signatures starts with the same sequence: `FF D8 FF`. Checking if the image file starts with it should be enough for our case.

jacekbogdanski reviewed (changes requested):
Works pretty nice! I see that this code had already some partial review in original PR, so there is not much left. For now, I've tested it with jpg, png, and gif. Please, see the review comments.

I also see that you marked this PR as a feature, although the original issue has been reported as a bug. I suppose that we could change it to feature as image support for OS X Word edition.

jacekbogdanski commented:
I also see that there are some jsduck errors 
![image](https://user-images.githubusercontent.com/10279586/97179677-63fea500-1799-11eb-918b-cc5d0c46d68c.png)

Please, update docs to match the correct types.

Comandeer commented:
> Please, update docs to match the correct types.

Hmm, that can be a little problematic, as [`Uint8Array`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array) is correct… Seems like jsduck is old enough to not recognize Typed Arrays :thinking:

Comandeer commented on plugins/pastetools/filter/image.js:
```diff
@@ -300,13 +362,60 @@
+				fileType = 'image/png';
+				break;
+			case '47494638':
+				fileType = 'image/gif';
```
Wow, I didn't know that Word supports GIFs :sweat\_smile: I was pretty sure they are converted to PNGs.

Comandeer replied:
Actually they are converted to PNG… So it works because [most browsers support APNG](https://caniuse.com/apng).

Safari in such a case use GIF format. It's strange that it isn't rendered correctly as we have GIF recognition :thinking:

Comandeer replied:
Adding GIF mime type to array of supported image formats seems to do the trick.

Comandeer commented:
Ok,

*   support for GIFs added by adding them to the array of supported image types,
*   tests with real images added,
*   I've also added throwing error when the image in blob is in unsupported format (reused the error introduced in #4291,
*   some refactoring.

jacekbogdanski commented on tests/plugins/pastefromword/_helpers/blob.js:
```diff
@@ -0,0 +1,113 @@
+	}
+
+	function simulatePasteBlob( editor, assertion, options ) {
+		var image = options && options.image || SAMPLE_PNG,
+			imageDataUrl = options && options.imageDataUrl || SAMPLE_PNG,
+			url = getObjectUrl( image, options && options.imageType ),
+			template = options && options.template || '<p{%CLASS%}>Hello<img style="height:200px; width:200px" src="{%URL%}" />world.</p>',
+			input = template.replace( /{%URL%}/g, url ).replace( /{%CLASS%}/g, ' class="MsoNormal"' ),
+			expected = template.replace( /{%URL%}/g, imageDataUrl ).replace( /{%CLASS%}/g, '' );
```
The code is a bit unreadable with all default option values. Maybe we could simplify it with some merging logic?

```js
CKEDITOR.tools.object.merge( defaultOptions, options || {} );
```

jacekbogdanski commented on plugins/pastetools/filter/image.js:
```diff
@@ -300,13 +390,50 @@
+		return new CKEDITOR.tools.promise( function( resolve ) {
+			CKEDITOR.ajax.load( blobUrlSrc, function( arrayBuffer ) {
+				var data = new Uint8Array( arrayBuffer ),
+					imageType = getImageTypeFromSignature( data.subarray( 0, 4 ) ),
```
Shouldn't we pass here the whole `arrayBuffer`? It seems like taking the first 4 characters is already done in `getImageTypeFromSignature` function? In that case, this should be probably renamed again to just `getImageType`.

jacekbogdanski reviewed (changes requested):
> Hmm, that can be a little problematic, as Uint8Array is correct… Seems like jsduck is old enough to not recognize Typed Arrays 🤔

I see that `Blob` is correctly recognized, so it's indeed weird that jsduck is not able to read this type. Maybe there is some workaround for custom type? If not, I suppose we will have to rename it into `Object` to avoid building warnings? Although, it's pity going some weird workarounds because the tooling doesn't work correctly.

Currently, ajax methods are a bit mixed up. Previously, we had:
* `load` - loading with `text` response
* `loadXml` - loading with `xml` response

now we have:
* `load` - by default text, but also supports `xml` and `arraybuffer`
* `loadXml` - still supporting only `xml`

Maybe, we should add explicit `loadBinaryData/loadData`? It probably doesn't really make sense deprecating `loadXml` as explicitness of this method is actually nice. Most readable seems to have `load` (without default type), `loadText`, `loadXml`, `loadBinaryData` but it's a bit too late for that.

jacekbogdanski commented:
Could you also rebase PR onto the latest major? Seems like there are some conficts.

Comandeer commented on plugins/pastetools/filter/image.js:
```diff
@@ -300,13 +390,50 @@
+	function convertBlobUrlToBase64( blobUrlSrc ) {
+		return new CKEDITOR.tools.promise( function( resolve ) {
+			CKEDITOR.ajax.load( blobUrlSrc, function( arrayBuffer ) {
+				var data = new Uint8Array( arrayBuffer ),
```
Yes, in old IEs this code is just ignored.

Comandeer commented:
Ok, I've addressed issues found during the review.

> Maybe, we should add explicit `loadBinaryData/loadData`? It probably doesn't really make sense deprecating `loadXml` as explicitness of this method is actually nice. Most readable seems to have `load` (without default type), `loadText`, `loadXml`, `loadBinaryData` but it's a bit too late for that.

TBH I don't find it's too late. We can always have additional methods and `load` with a default response type.

I'll raise an issue for it.

jacekbogdanski reviewed (approved):
Good job!
